Init error tracking guides and concepts (#12404)
PostHog/posthog.com: contents/docs/error-tracking/fingerprints.mdx

---
title: Fingerprints
---

Every captured exception is assigned a fingerprint. This fingerprint is used to group similar exceptions into issues. Fingerprints can also be manually set during capture. This page covers how fingerprints are generated and how they're used.

## Fingerprint and issue grouping

> Every exception has a fingerprint, whether generated or defined by the user. Each fingerprint links to exactly one issue. Exceptions that share the same fingerprint are grouped together.

Reviewer: ```suggestion
Every exception has a fingerprint, whether generated or defined by the user. Each fingerprint links to exactly one issue. Exceptions that share the same fingerprint define an issue.
```

Multiple different fingerprints can point to the same issue (a many-to-one relationship) if you [merge issues](/docs/error-tracking/managing-issues#merging-issues). 

## How are fingerprints generated?

Fingerprints are built iteratively using components of the exception event. The flowchart below shows how fingerprints are generated.

```mermaid
flowchart LR
    A[Add exception type to fingerprint] --> B{Stack trace<br/>available?}
    
    B -->|No| C[Add error message to fingerprint]
    
    C --> D[Final fingerprint]
    
    B -->|Yes| E{In-app frames<br/>exist?}
    
    E -->|No| F[Add first frame to fingerprint]
    
    E -->|Yes| G[Add in-app frame to fingerprint<br/>Priority: resolved > unresolved]
    
    F --> D
    G --> D
```

<details>
<summary>The flowchart in text</summary>

Fingerprints are generated by considering the following in combination:
1. The exception type
2. If there's no resolved stack trace, add the error message to the fingerprint
3. If there are stack traces
  1. If there are no in-app stack frames (frames from your code, not a dependency), use the first frame of the stack trace

>   2. Otherwise only use a frame if it's in app, **and** is either resolved or no other frames are resolved and available

Reviewer: This is an excellently concise description of an otherwise complex flow 😂

Reviewer 2: @daibhin @gewenyu99  updated again because nested order list items don't work :( 

> 1. The exception type
> 2. If there's no resolved stack trace, add the error message to the fingerprint
> 3. If there are stack traces but no in-app frames (frames from your code, not a dependency), use the first frame > of the stack trace
> 4. If there are stack traces, in-app frames, and source maps are available, use the resolved in-app stack frames
> 5. If there are stack traces, in-app frames, and source maps are *not* available, use the first in-app stack frame

</details>

If there are multiple exceptions leading to one another, a single fingerprint will be generated for that chain of exceptions.

### Ensuring accurate fingerprints

> [Resolved stack traces](/docs/error-tracking/stack-traces) are critical for accurate fingerprinting. Without accurate stack traces, PostHog cannot group exceptions consistently. If you have not uploaded source maps, you should do so following this [guide](/docs/error-tracking/upload-source-maps).

Reviewer 3: ```suggestion
[Resolved stack traces](/docs/error-tracking/stack-traces) are critical for accurate fingerprinting. Without accurate stack traces, PostHog cannot group exceptions consistently. If you have not uploaded source maps, follow the [source map guide](/docs/error-tracking/upload-source-maps) to do so.
```
